Reorder defines so xlC16 finds the xlc before gnu

DanHeidinga opened the pull request:
xlC16 appears to be based on clang and now recognizes the
__GNUC__ defines.  Reorder this code so that the compiler
sees the __xlC__-specific case first.

+ additional changes to support xlc16

issue: #3775

Signed-off-by: Dan Heidinga <daniel_heidinga@ca.ibm.com>

DanHeidinga commented:
Marking as WIP until we've confirmed that the xlc16 succeeds with this change

DanHeidinga commented:
@vsebe I've pushed a new commit that hopefully works around the missing support for `pragma mc_func`

charliegracie commented:
@genie-omr build plinux aix

DanHeidinga commented:
This PR resolves the issues listed so far.  There may be other issues lurking still with xlc16 but I'd like to land this while we keep investigating

Leonardo2718 commented on include_core/AtomicSupport.hpp:
```diff
@@ -410,7 +413,7 @@ class VM_AtomicSupport
 #elif defined(__xlC__) /* defined(J9ZOS390) */
```
Should this guard have been removed (replaced by the one at line 399)?

DanHeidinga commented:
Thanks @Leonardo2718.  I've updated the copyright date and deleted the duplicate case.  There was also some spaces vs tabs issues that I corrected at the same time.

Leonardo2718 commented:
@genie-omr build all
